File: DataStructures/Trees/BFS・DFS/leetcode/39. Combination Sum/GPT/Combination-Sum.py
# ```python
# Runtime
# 3
# ms
# Beats
# 97.48%
# Memory
# 17.74
# MB
# Beats
# 87.72%

from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        """
        与えられた候補数字(candidates)から、合計がtargetとなる全ての組み合わせを返す。
        同じ数字は何度でも利用可能で、順序の異なる同一組み合わせは重複しない。

        Parameters
        ----------
        candidates : List[int]
            候補となる整数のリスト（全て異なる）
        target : int
            合計値（目標値）

        Returns
        -------
        List[List[int]]
            targetに一致する組み合わせのリスト

        制約
        ----
        1 <= len(candidates) <= 30
        2 <= candidates[i] <= 40
        1 <= target <= 40

        計算量
        -----
        時間計算量: O(N^(T/M)) （N=候補数, T=target, M=候補の最小値）
        空間計算量: O(T/M)（再帰スタック）+ O(#解答)（結果保存）
        """
        res: List[List[int]] = []
        path: List[int] = []

        def dfs(start: int, current_sum: int) -> None:
            # targetに到達 → 結果追加
            if current_sum == target:
                res.append(path.copy())
                return
            # 候補を順に試す
            for i in range(start, len(candidates)):
                next_sum = current_sum + candidates[i]
                # 枝刈り: targetを超えたら終了
                if next_sum > target:
                    continue
                path.append(candidates[i])
                dfs(i, next_sum)  # 同じ数字を再利用可能
                path.pop()  # 戻って他の候補を試す

        dfs(0, 0)
        return res
# ```

# ---

# ### 実装のポイント

# 1. **`start`インデックスで順序違いの重複を回避**

#    * 例: \[2,3,2] と \[2,2,3] は同じ組み合わせなので1度だけ返す
# 2. **枝刈り (`if next_sum > target`)**

#    * 計算量を削減し、高速化
# 3. **`path.copy()`で結果保存**

#    * `path`は参照なので、コピーしてから保存しないと変更が影響してしまう
# 4. **Pylance型エラー防止**

#    * `List[int]`, `List[List[int]]` を明示

# ---

# このコードは

# * LeetCodeの制約（最大150通りの組み合わせ）で **TLEやメモリ超過なし**
# * Python (CPython 3.11.4) で安全に動作
# * Pylanceの型検査でもエラーなし

# ---

# もし希望があれば、このPython版の処理も **探索ツリー図** を付けて詳しく解析できます。
# これをやると理解がさらに深まります。

File: DataStructures/Trees/BFS・DFS/leetcode/39. Combination Sum/GPT/Combination-Sum.js
// ```javascript
Runtime;
1;
ms;
Beats;
97.55 % Memory;
59.64;
MB;
Beats;
65.89 %
    /**
     * Function: combinationSum
     * -------------------------
     * Returns all unique combinations of numbers from `candidates` that sum up to `target`.
     * Each number from `candidates` can be used an unlimited number of times.
     *
     * @param {number[]} candidates - Distinct integers available for combination (length ≤ 30, 2 ≤ value ≤ 40)
     * @param {number} target - Target sum to achieve (1 ≤ target ≤ 40)
     * @returns {number[][]} - Array of unique combinations where sum equals target
     *
     * Time Complexity: O(N^(T/M)) in the worst case
     *   - N = number of candidates
     *   - T = target value
     *   - M = smallest candidate value
     *   (Efficient pruning is used to avoid unnecessary recursion)
     * Memory Usage: O(T/M) for recursion stack + O(#combinations) for results
     *   - Given constraints, memory remains low
     */
    function combinationSumJs(candidates, target) {
        const res = [];
        const path = [];

        /**
         * DFS (backtracking) helper function
         * @param {number} start - Current index in candidates to consider
         * @param {number} sum - Current sum of the chosen numbers
         */
        function dfs(start, sum) {
            // If we reach the target, store the current combination
            if (sum === target) {
                res.push([...path]);
                return;
            }

            // Loop through candidates starting from `start` to avoid duplicates
            for (let i = start; i < candidates.length; i++) {
                const nextSum = sum + candidates[i];
                // Prune branches where sum exceeds target
                if (nextSum > target) continue;

                path.push(candidates[i]); // Choose candidate
                dfs(i, nextSum); // Allow reuse of same number
                path.pop(); // Backtrack
            }
        }

        dfs(0, 0);
        return res;
    };
// ```

// この実装は

// * **再帰的DFS（深さ優先探索）＋バックトラッキング** を用いて全探索
// * **ソート不要**（ただし、ソートすれば更に枝刈りが効く場合あり）
// * **メモリ効率**は、`path`と`res`のみ保持し、再帰の深さも`target / min(candidates)`程度

// これならLeetCode制約（最大150通りの組み合わせ）でも十分高速に動作します。
